### PR TITLE
Implement enhancement #106 -- VAE with disentangled latent representation of a qc matrix

### PR DESCRIPTION
**Pull request by jstjohn**
## New features:
* This implements the model discussed in issue #106.
* Update the command line `run_benchmarks.py` to take the `model` as an argument.
* Adds in a one-hot encoded representation of the batch indices, along with the sample mean/max as an example of qc data that could be passed in per sample.

### CODE REVIEW

**Comment by jeff-regier** [comment]
Hi John! Thanks a lot for this contribution. I'll start looking over it carefully---it may take some thought to figure out how to integrate this code into our code base.

@romain-lopez and @Edouard360 will you look over it to and post your thoughts? We hadn't really discussed adding the HSIC penalty to our PyTorch codebase. If we're going to add it, we probably need to include 1) examples of how to use it in notebooks, and 2) documentation of it.

File: scvi/metrics/log_likelihood.py [hunk 1]
```diff
@@ -14,17 +15,49 @@ def compute_log_likelihood(vae, data_loader):
...
+                                          batch_index=batch_index, y=labels)
         log_lkl += torch.sum(reconst_loss).item()
     n_samples = (len(data_loader.dataset)
                  if not (hasattr(data_loader, 'sampler') and hasattr(data_loader.sampler, 'indices')) else
                  len(data_loader.sampler.indices))
     return log_lkl / n_samples
 
 
+def sigmoid_cross_entropy_with_logits(x, z):
+    return torch.clamp(x, min=0.0) - x * z + torch.log(1 + torch.exp(-torch.abs(x)))
```
Comment (jeff-regier) [inline scvi/metrics/log_likelihood.py]: Is there a numerically stable version of this already implemented in PyTorch? There's some discussion of it here: https://discuss.pytorch.org/t/equivalent-of-tensorflows-sigmoid-cross-entropy-with-logits-in-pytorch/1985

File: scvi/models/svaec_qc.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+from scvi.models.utils import one_hot, enumerate_discrete
+
+torch.backends.cudnn.benchmark = True
+
+
+# VAE model
+class SVAECQC(nn.Module):
+    def __init__(self, n_input, n_input_qc, n_hidden=128, n_latent=10, n_latent_qc=3, n_layers=1, dropout_rate=0.1,
+                 dispersion="gene", log_variational=True, reconstruction_loss="zinb", n_batch=0, n_labels=0,
+                 n_s_decoder=25, use_cuda=False, y_prior=None):
```
Comment (jeff-regier) [inline scvi/models/svaec_qc.py]: Is a lot of this code copied from another module? If so, would it make sense to subclass that module rather than duplicating the code?

Comment (jeff-regier) [reply scvi/models/svaec_qc.py]: @Edouard360 -- Glad for your thoughts on that too.

File: scvi/models/svaec_qc.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        if self.log_variational:
+            x_ = torch.log(1 + x_)
+
+        # Sampling
+        qz1_m, qz1_v, z1 = self.z_encoder(x_)
+        qu_m, qu_v, u = self.u_encoder(torch.cat((x_, qc_), 1))
+        ql_m, ql_v, library = self.l_encoder(x_)
+
+        # Enumerate choices of label
+        if not is_labelled:
```
Comment (Edouard360) [inline scvi/models/svaec_qc.py]: It gets a little messy for the semi-supervised models:
I had done locally a `broadcast_labels` function, that writes:
```
        ys, xs, batch_index, local_l_var, local_l_mean, qz1_m, qz1_v, z1 = (
            broadcast_labels(
                ys, xs, batch_index, local_l_var, local_l_mean, qz1_m, qz1_v, z1,n_broadcast=self.n_labels
            )
        )
```
It's a little prettier than all those copy paste lines.. I will be pushing that

File: scvi/models/svaec_qc.py (hunk 2 above)
Comment (Edouard360) [reply scvi/models/svaec_qc.py]: I think subclassing would be easy to implement. The __init__ and forward function only differ in a few lines.

File: scvi/dataset/dataset.py [hunk 4]
```diff
@@ -30,7 +30,26 @@ def __init__(self, X, local_means, local_vars, batch_indices, labels, gene_names
...
         self.batch_indices = batch_indices
         self.n_batches = len(np.unique(batch_indices))
-        self.dense = type(self.X) is np.ndarray
+        self.dense = isinstance(self.X, np.ndarray)
+        if qc is None:
+            onehot_encoder = OneHotEncoder(sparse=(not self.dense))
+            onehot_batches = onehot_encoder.fit_transform(self.batch_indices)
+            sample_means = np.mean(self.X, axis=1)
+            sample_max = np.max(self.X, axis=1)
+            qc_parts = (onehot_batches, sample_means, sample_max)
```
Comment (Edouard360) [inline scvi/dataset/dataset.py]: These quality control metrics are interesting.. I see the normalization in [0,1] with the MinMaxScaler and the modelling of the variable as a bernouilli. @romain-lopez : Is it something we should systematically take as qc metrics ? From your paper, you mention 5 different factors but they come from meta information about each barcodes from the experiment. This proposition looks very interesting to me.

Comment (Edouard360) [reply scvi/dataset/dataset.py]: however, just as a slight note there is no need to OneHotEncode the batch_indices here since it is anyway done in the training loop

Comment (romain-lopez) [reply scvi/dataset/dataset.py]: In the NIPS submission, I consider only the QC that are related to the alignment. These ones are often proportions. Sometime other however (proxy for amplification, proxy for sampling) are positive (exponential, poisson, gamma...).
And I agree, I condense this barcode information into cell information since our graphical model does not include the transcripts (unlike the LDA).

**Comment by romain-lopez** [comment]
Thanks @jstjohn for this contribution ! As Jeffrey underlined, it would be great to have a notebook with minimal code to run that and how to use it.

More insight, maybe it would also be better to specify for each dimension of the QC matrix a link function (logit for proportion, log for counts) and systematically use a normal variable for the modeling with the mean constrained with the right function ? That way the user is free to do however he would like to ?

**Comment by jeff-regier** [comment]
@jstjohn We discussed this PR at our last weekly project meeting. The consensus seems to be that we'll need do a lot more testing and experimentation with the HCV algorithm before merging the code--and not just unit testing, but testing it with a variety of datasets, and a variety of QC matrices. Romain is away this summer so I don't expect that will happen before September or so. In the meantime, if you've been trying this code with your own datasets, we'd certainly appreciate hearing about your experiences.
